File: KCD/HMKCDeck+Extension.m
//
//  HMKCDeck+Extension.m
//  KCD
//
//  Created by Hori,Masaki on 2014/10/05.
//  Copyright (c) 2014年 Hori,Masaki. All rights reserved.
//

#import "HMKCDeck+Extension.h"

@implementation HMKCDeck (Extension)

- (HMKCShipObject *)shipOfShipNumber:(NSInteger)shipNumber
{
	id ship = nil;
	NSError *error = nil;
	NSManagedObjectContext *moc = [self managedObjectContext];
	NSString *key = [NSString stringWithFormat:@"ship_%ld", shipNumber];
	[self willAccessValueForKey:key];
	NSNumber *shipIdNumber = [self valueForKey:key];
	[self didAccessValueForKey:key];
	NSInteger shipId = [shipIdNumber integerValue];
	NSArray *array = nil;
	if(shipId != -1) {
		NSFetchRequest *request = [NSFetchRequest fetchRequestWithEntityName:@"Ship"];
		NSPredicate *predicate = [NSPredicate predicateWithFormat:@"id = %ld", shipId];
		[request setPredicate:predicate];
		array = [moc executeFetchRequest:request error:&error];
	}
	if(shipId != -1 && array.count == 0) {
		NSLog(@"Could not found ship of id %@", shipIdNumber);
	} else {
		ship = array[0];
	}
	
	return ship;
}
- (HMKCShipObject *)flagShip
{
	return [self shipOfShipNumber:0];
}
- (HMKCShipObject *)secondShip
{
	return [self shipOfShipNumber:1];
}
- (HMKCShipObject *)thirdShip
{
	return [self shipOfShipNumber:2];
}
- (HMKCShipObject *)fourthShip
{
	return [self shipOfShipNumber:3];
}
- (HMKCShipObject *)fifthShip
{
	return [self shipOfShipNumber:4];
}
- (HMKCShipObject *)sixthShip
{
	return [self shipOfShipNumber:5];
}

	

@end

File: KCD/HMKCDeck+Extension.swift
//
//  HMKCDeck.swift
//  KCD
//
//  Created by Hori,Masaki on 2014/12/31.
//  Copyright (c) 2014年 Hori,Masaki. All rights reserved.
//

import Foundation
import CoreData

extension HMKCDeck {
	private func shipOfShipNumber(shipNumber: Int) -> HMKCShipObject? {
		let key = "ship_\(shipNumber)"
		willAccessValueForKey(key)
		let shipID = valueForKey(key) as Int
		didAccessValueForKey(key)
		var array: [HMKCShipObject]? = nil
		if shipID == -1 { return nil }
		let request = NSFetchRequest(entityName: "Ship")
		request.predicate = NSPredicate(format: "id = %ld", argumentArray: [shipID])
		var error: NSError? = nil
		array = managedObjectContext?.executeFetchRequest(request, error: &error) as? [HMKCShipObject]
		if array == nil {
			println(__FUNCTION__, ": Could not found ship of id \(shipID)")
			if error != nil {
				println("\terror -> \(error!)")
			}
		}
		
		return array?[0]
	}
	
	var flagShip: HMKCShipObject? {
		return shipOfShipNumber(0)
	}
	var secondShip: HMKCShipObject? {
		return shipOfShipNumber(1)
	}
	var thirdShip: HMKCShipObject? {
		return shipOfShipNumber(2)
	}
	var fourthShip: HMKCShipObject? {
		return shipOfShipNumber(3)
	}
	var fifthShip: HMKCShipObject? {
		return shipOfShipNumber(4)
	}
	var sixthShip: HMKCShipObject? {
		return shipOfShipNumber(5)
	}
}
